src/async-wrap.cc

```diff
@@ -13,16 +14,91 @@ using v8::Function;
 using v8::FunctionCallbackInfo;
 using v8::Handle;
 using v8::HandleScope;
+using v8::HeapProfiler;
 using v8::Integer;
 using v8::Isolate;
 using v8::Local;
 using v8::Object;
+using v8::RetainedObjectInfo;
 using v8::TryCatch;
 using v8::Value;
 using v8::kExternalUint32Array;
 
 namespace node {
 
+static const char* const provider_names[] = {
+#define V(PROVIDER)                                                           \
+  #PROVIDER,
+  NODE_ASYNC_PROVIDER_TYPES(V)
+#undef V
+};
+
+
+class RetainedAsyncInfo: public RetainedObjectInfo {
+ public:
+  explicit RetainedAsyncInfo(uint16_t class_id, Local<Value> wrapper);
+
+  virtual void Dispose() override;
+  virtual bool IsEquivalent(RetainedObjectInfo* other) override;
+  virtual intptr_t GetHash() override;
+  virtual const char* GetLabel() override;
+  virtual intptr_t GetSizeInBytes() override;
+
+ private:
+  const char* label_;
+  AsyncWrap* wrap_;
+  int length_;
+};
+
+
+RetainedAsyncInfo::RetainedAsyncInfo(uint16_t class_id, Local<Value> wrapper)
+    : label_(provider_names[class_id - ASYNC_ID_OFFSET]),
+      wrap_(nullptr),
+      length_(0) {
+  CHECK_NE(0xA1C, class_id);
+  CHECK(wrapper->IsObject());
+  CHECK(!wrapper.IsEmpty());
+  Local<Object> object = wrapper.As<Object>();
+  wrap_ = Unwrap<AsyncWrap>(object);
+  CHECK_NE(nullptr, wrap_);
+  length_ = wrap_->self_size();
+}
```

Comment:
I would move the checks and the cast into WrapperInfo() and change the constructor's prototype to take a `AsyncWrap*`.  That way, you can assign `wrap_` and `length_` with `wrap_(wrap), length_(wrap->self_size())`.  Maybe make them `const` in the process.